File: typescript/packages/core/src/team/watch.ts
import type { MemberRef } from "../log";
import {
  addressed,
  type CallContext,
  callerOf,
  callMailId,
  callRequest,
  decide,
  decision,
  isRefusal,
  named,
  type Refused,
  recorded,
} from "./call";
import {
  type CloseContext,
  finishWait,
  publicResult,
  type Waiting,
} from "./close";
import { TEAM_CONSTANTS } from "./constants";
import { parkCall } from "./park";
import type { Request, Target } from "./request";
import type { MonitorResult, Waited, Wire } from "./results";
import { type MemberRow, refOf, settledOf } from "./rows";

// The model tools wait and monitor (spec/schema/README.md, "Teams", Waits and monitors; design
// §4.12 and §4.13). Each reads its targets' rows in its own transaction: a settled target is
// recorded as member_observed (its committed result, copied), any other gets a monitor row that
// its idle or end append fires. Never both, never neither. Reference:
// spec/tools/fixtures/ops_observe.py.

const SETTLED: ReadonlySet<MemberRow["state"]> = new Set(["idle", "ended"]);

/** member_observed: the target's committed result, and where it is. */
async function observe(
  ctx: Pick<CloseContext, "tx" | "batch">,
  monitorId: string,
  row: MemberRow,
): Promise<void> {
  if (row.branch_id === null) throw new Error("a settled member has a branch");
  const { result, seq } = await settledOf(ctx.tx, row);
  ctx.batch.add({
    type: "member_observed",
    type_version: 1,
    critical: true,
    actor: { kind: "host" },
    data: {
      monitor_id: monitorId,
      result,
      source: { thread_id: row.thread_id, branch_id: row.branch_id, seq },
    },
  });
}

/** How many listed members must settle: all, any (one), or a count. */
export type WaitMode = "all" | "any" | number;

/**
 * The model's wait (mode all, the default deadline): the listed members, a repeat dropped, then
 * the wait. A re-dispatched wait only parks, and so does one left waiting.
 */
export async function wait(
  ctx: CallContext,
  args: { readonly members: readonly string[] },
  timeoutMs: number = TEAM_CONSTANTS.askWaitDefaultMs,
): Promise<Wire<Waited> | Waiting | Refused> {
  const caller = await callerOf(ctx);
  if (caller === undefined) throw new Error("a team tool call outside a team");
  const waitId = callMailId(ctx);
  const again = ctx.chain.events.some(
    (l) =>
      l.kind === "event" &&
      l.event.type === "wait_started" &&
      l.event.data.wait_id === waitId,
  );
  if (!again) {
    const targets = [...new Set(args.members)].map((m) => named(ctx, m));
    const got = await openWait(await callRequest(ctx), closing(ctx), targets, {
      mode: "all",
      timeoutMs,
    });
    if (got.status !== "waiting") return got;
  }
  await parkCall(ctx, caller);
  return { status: "waiting", wait_id: waitId };
}

/**
 * An operator wait's members, a repeat dropped (they are frozen at the call); invalid_request for
 * an empty list, or a numeric mode below 1 or above their count, refused before any writer is
 * taken.
 */
export function waitMembers(
  members: readonly MemberRef[],
  mode: WaitMode | undefined,
): readonly MemberRef[] | "invalid_request" {
  const key = (m: MemberRef) =>
    JSON.stringify([m.tenant, m.team, m.name, m.generation]);
  const distinct = [...new Map(members.map((m) => [key(m), m])).values()];
  const bad =
    distinct.length === 0 ||
    (typeof mode === "number" && (mode < 1 || mode > distinct.length));
  return bad ? "invalid_request" : distinct;
}

/**
 * A wait, for a model call or an operator request: each target known at its generation, then one
 * monitor decision each; wait_started, an observation per settled member, and the finish when
 * that meets the mode. Returns what it recorded, or waiting.
 */
export async function openWait(
  req: Request,
  close: CloseContext,
  targets: readonly Target[],
  how: { readonly mode: WaitMode; readonly timeoutMs: number },
): Promise<Wire<Waited> | Waiting | Refused> {
  const rows: MemberRow[] = [];
  for (const target of targets) {
    const row = await target.row();
    if ("refused" in row) return req.refuse(row);
    rows.push(row);
  }
  for (const row of rows) {
    const denied = req.decide("monitor", row.name);
    if (denied !== undefined) return req.refuse(denied);
  }
  const cap = TEAM_CONSTANTS.askWaitDefaultMs;
  const started = req.batch.add({
    type: "wait_started",
    type_version: 1,
    critical: true,
    actor: { kind: "host" },
    data: {
      wait_id: req.mailId,
      members: rows.map((r) => refOf(req.team, r)),
      mode: how.mode,
      deadline: req.batch.now + Math.min(how.timeoutMs, cap),
    },
  });
  for (const row of rows)
    if (SETTLED.has(row.state))
      await observe(close, `${close.branchId}:${started}:${row.name}`, row);
  return await finishWait(close, req.mailId, { deadline: false });
}

/** The call's append as a close: the caller's own thread and branch. */
function closing(ctx: CallContext): Parameters<typeof finishWait>[0] {
  return {
    ...ctx,
    threadId: ctx.call.thread_id,
    branchId: ctx.call.branch_id,
  };
}

/**
 * monitor: its decision, then the member; monitor_set, and for an ended target its observation
 * and the ended result at once, else an end monitor row whose firing opens a turn when idle.
 */
export async function monitor(
  ctx: CallContext,
  args: { readonly member: string },
): Promise<Wire<MonitorResult> | Refused> {
  const caller = await callerOf(ctx);
  if (caller === undefined) throw new Error("a team tool call outside a team");
  const denied = decide(
    ctx,
    "monitor",
    args.member,
    decision(ctx, caller, "monitor", args.member),
  );
  if (denied !== undefined) return recorded(ctx, denied);
  const row = await addressed(ctx, caller, args.member);
  if (isRefusal(row)) return recorded(ctx, row);
  const member = refOf(caller.team, row);
  const set = ctx.batch.add({
    type: "monitor_set",
    type_version: 1,
    critical: true,
    actor: { kind: "host" },
    data: { member },
  });
  if (row.state !== "ended")
    return recorded(ctx, {
      member,
      status: "monitoring",
    } satisfies Wire<MonitorResult>);
  await observe(ctx, `${ctx.call.branch_id}:${set}:${row.name}`, row);
  const { result } = await settledOf(ctx.tx, row);
  return recorded(ctx, {
    result: await publicResult(result, ctx.read),
    status: "ended",
  } satisfies Wire<MonitorResult>);
}

File: python/src/threads/team/watch.py
"""The model tools wait and monitor (spec/schema/README.md, "Teams", Waits and monitors; design
§4.12 and §4.13). Each reads its targets' rows in its own transaction: a settled target is recorded
as member_observed (its committed result, copied), any other gets a monitor row that its idle or
end append fires. Never both, never neither. Reference: spec/tools/fixtures/ops_observe.py."""

from collections.abc import Sequence
from typing import Literal

from pydantic import JsonValue

from threads.log import MemberRef, WaitStartedEvent
from threads.pos_int import is_pos_int
from threads.store.lines import Draft
from threads.team.call import (
    CallContext,
    Refusal,
    addressed,
    call_mail_id,
    call_request,
    caller_of,
    decide,
    decision,
    named,
    recorded,
)
from threads.team.close import CloseContext, finish_wait, public_result
from threads.team.constants import TEAM_CONSTANTS
from threads.team.park import park_call
from threads.team.request import Request, Target
from threads.team.rows import MemberRow, ref_of, settled_of

_SETTLED = frozenset({"idle", "ended"})


def _observe(ctx: CallContext | CloseContext, monitor_id: str, row: MemberRow) -> JsonValue:
    """member_observed: the target's committed result, and where it is."""
    if row.branch_id is None:
        raise AssertionError("a settled member has a branch")
    settled = settled_of(ctx.conn, row)
    source: dict[str, JsonValue] = {
        "thread_id": row.thread_id,
        "branch_id": row.branch_id,
        "seq": settled.seq,
    }
    data: dict[str, JsonValue] = {
        "monitor_id": monitor_id,
        "result": settled.result,
        "source": source,
    }
    ctx.batch.add(Draft("member_observed", data))
    return settled.result


def _closing(ctx: CallContext) -> CloseContext:
    """The call's append as a close: the caller's own thread and branch."""
    call = ctx.call
    return CloseContext(ctx.conn, ctx.batch, call.thread_id, call.branch_id, ctx.fold, ctx.read)


type WaitMode = Literal["all", "any"] | int
"""How many listed members must settle: all, any (one), or a count."""


def wait(
    ctx: CallContext,
    members: Sequence[str],
    timeout_ms: int = TEAM_CONSTANTS.ask_wait_default_ms,
) -> JsonValue:
    """The model's wait (mode all, the default deadline): the listed members, a repeat dropped,
    then the wait. A re-dispatched wait only parks, and so does one left waiting."""
    caller = caller_of(ctx)
    wait_id = call_mail_id(ctx)
    again = any(
        isinstance(e, WaitStartedEvent) and e.data.wait_id == wait_id for e in ctx.fold.events
    )
    if not again:
        targets = [named(ctx, m) for m in dict.fromkeys(members)]
        got = open_wait(call_request(ctx), _closing(ctx), targets, "all", timeout_ms)
        if not isinstance(got, dict) or got.get("status") != "waiting":
            return got
    park_call(ctx, caller.provenance)
    return {"status": "waiting", "wait_id": wait_id}


def wait_members(
    members: Sequence[MemberRef], mode: WaitMode | None
) -> tuple[MemberRef, ...] | Literal["invalid_request"]:
    """An operator wait's members, a repeat dropped (they are frozen at the call);
    invalid_request for an empty list, or a mode that is neither "all", "any", nor a positive
    integer no greater than their count, refused before any writer is taken."""
    distinct = tuple(dict.fromkeys(members))
    if not distinct:
        return "invalid_request"
    if mode is None:
        return distinct
    if isinstance(mode, str):
        return distinct if mode in ("all", "any") else "invalid_request"
    bad = not is_pos_int(mode) or mode > len(distinct)
    return "invalid_request" if bad else distinct


def open_wait(
    req: Request,
    close: CloseContext,
    targets: Sequence[Target],
    mode: WaitMode,
    timeout_ms: int,
) -> JsonValue:
    """A wait, for a model call or an operator request: each target known at its generation,
    then one monitor decision each; wait_started, an observation per settled member, and the
    finish when that meets the mode. Returns what it recorded, or waiting."""
    rows: list[MemberRow] = []
    for target in targets:
        row = target.row()
        if isinstance(row, Refusal):
            return req.refuse(row)
        rows.append(row)
    for row in rows:
        denied = req.decide("monitor", row.name)
        if denied is not None:
            return req.refuse(denied)
    cap = TEAM_CONSTANTS.ask_wait_default_ms
    started_data: dict[str, JsonValue] = {
        "wait_id": req.mail_id,
        "members": [ref_of(req.team, r) for r in rows],
        "mode": mode,
        "deadline": req.batch.now + min(timeout_ms, cap),
    }
    started = req.batch.add(Draft("wait_started", started_data))
    for row in rows:
        if row.state in _SETTLED:
            _observe(close, f"{close.branch_id}:{started}:{row.name}", row)
    return finish_wait(close, req.mail_id, cause=None, deadline=False)


def monitor(ctx: CallContext, member: str) -> JsonValue:
    """monitor: its decision, then the member; monitor_set, and for an ended target its
    observation and the ended result at once, else an end monitor row whose firing opens a turn
    when idle."""
    caller = caller_of(ctx)
    denied = decide(ctx, "monitor", member, decision(ctx, caller, "monitor", member))
    if denied is not None:
        return recorded(ctx, denied)
    row = addressed(ctx, caller, member)
    if isinstance(row, Refusal):
        return recorded(ctx, row)
    ref = ref_of(caller.team, row)
    set_ = ctx.batch.add(Draft("monitor_set", {"member": ref}))
    if row.state != "ended":
        return recorded(ctx, {"member": ref, "status": "monitoring"})
    result = _observe(ctx, f"{ctx.call.branch_id}:{set_}:{row.name}", row)
    return recorded(ctx, {"result": public_result(result, ctx.read), "status": "ended"})
